cmd/virt-apiservice-register/virt-apiservice-register.go

```diff
@@ -0,0 +1,737 @@
+	)
+
+	// Parse the command line:
+	flag.Parse()
+
+	// Check that mandatory options have been provided:
+	ok := true
+	if apiGroup == "" {
+		glog.Errorf("The -api-group option is mandatory")
+		ok = false
+	}
+	if apiVersion == "" {
+		glog.Errorf("The -api-version option is mandatory")
+		ok = false
+	}
+	if targetSelector == "" {
+		glog.Errorf("The -target-selector option is mandatory")
+		ok = false
+	}
+	if !ok {
+		os.Exit(1)
+	}
+
+	// Calculate default names for the services and the secret:
+	dotReplacer := strings.NewReplacer(".", "-")
+	defaultName := fmt.Sprintf(
+		"apiservice-%s-%s",
+		dotReplacer.Replace(apiGroup),
+		dotReplacer.Replace(apiVersion),
+	)
+	if serviceName == "" {
+		serviceName = defaultName
+	}
+	if secretName == "" {
+		secretName = defaultName
+	}
+
+	// Run the tool and report unhandled errors:
+	err := run()
+	if err != nil {
+		reportError(err)
+		os.Exit(1)
+	} else {
+		os.Exit(0)
+	}
+}
+
+func run() error {
+	// Load the configuration and create the API clients. Note that in addition to the typed client
+	// we also need a dynamic client, because the 'client-go' package doesn't currently support the
+	// API service registration API in the typed way.
+	client, dyclient, err := createClients()
+	if err != nil {
+		return err
+	}
+
+	// Find the CA, as we will need it to create the secret and the API service:
+	ca, err := findCA(client)
+	if err != nil {
+		return err
+	}
+
+	// Make sure that the secret exists:
+	err = ensureSecret(client, ca)
+	if err != nil {
+		return err
+	}
+
+	// Make sure that the service exists:
+	err = ensureService(client)
+	if err != nil {
+		return err
+	}
+
+	// Make sure that the API service exists:
+	err = ensureAPIService(dyclient, ca)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// createClients loads the Kubernetes API client configuration using the value of the kubeconfig
+// command line option and creates the API clients.
+//
+func createClients() (client *kubernetes.Clientset, dyclient *dynamic.Client, err error) {
+	// Create the typed client, which will be used for most of the interactions with the API:
+	cfg, err := clientcmd.BuildConfigFromFlags("", kubeConfig)
+	if err != nil {
+		return
+	}
+	client, err = kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return
+	}
+
+	// Create the dynamic client, which will be used for the API service registration API, as it
+	// isn't yet supported by the typed part of the 'client-go' package.
+	dycfg, err := clientcmd.BuildConfigFromFlags("", kubeConfig)
+	if err != nil {
+		return
+	}
+	dycfg.APIPath = "/apis/"
+	dycfg.GroupVersion = &schema.GroupVersion{
+		Group:   "apiregistration.k8s.io",
+		Version: "v1beta1",
+	}
+	dyclient, err = dynamic.NewClient(dycfg)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+// ensureSecret ensures that the secret containing the CA, private key and certificate exists. If it
+// doesn't exist it will be created.
```

Comment:
This comment says the secret will be created if it doesn't exist. However, in the code will exit with an error if the secret does not exist without attempting to create the secret.